src/p2pMessage.ts

```diff
@@ -0,0 +1,300 @@
+
+        switch (protocolId) {
+            case MessageType.REQUEST_ID: {
+                const extensionName = decodedbytes[COMMON];
+                const extensionVersion = decodedbytes[COMMON + 1];
+
+                return new NegotiationMessage(version, seq, {
+                    type: "request",
+                    extensionName,
+                    extensionVersion
+                });
+            }
+            case MessageType.ALLOWED_ID:
+                return new NegotiationMessage(version, seq, {
+                    type: "allowed",
+                    version: decodedbytes[COMMON].readUIntBE(0, 1)
+                });
+            case MessageType.DENIED_ID:
+                return new NegotiationMessage(version, seq, { type: "denied" });
+            default:
+                throw Error("Unreachable");
+        }
+    }
+}
+
+type NegotiationBody =
+    | INegotiationRequest
+    | INegotiationAllowed
+    | INegotiationDenied;
+
+interface INegotiationRequest {
+    type: "request";
+    extensionName: string;
+    extensionVersion: Array<number>;
+}
+
+interface INegotiationAllowed {
+    type: "allowed";
+    version: number;
+}
+
+interface INegotiationDenied {
+    type: "denied";
+}
+
+export class ExtensionMessage {
+    private version: number;
+    private extensionName: string;
+    private extensionVersion: number;
+    private data: IData;
+
+    constructor(
+        version: number,
+        extensionName: string,
+        extensionVersion: number,
+        data: IData
+    ) {
+        this.version = version;
+        this.extensionName = extensionName;
+        this.extensionVersion = extensionVersion;
+        this.data = data;
+    }
+
+    protocolId(): number {
+        switch (this.data.type) {
+            case "encrypted":
+                return MessageType.ENCRYPTED_ID;
+            case "unencrypted":
+                return MessageType.UNENCRYPTED_ID;
+        }
+    }
+
+    toEncdoeObject(): Array<any> | number | string {
+        return [
+            this.version,
+            this.protocolId(),
+            this.extensionName,
+            this.extensionVersion,
+            `0x${this.data.data.toString("hex")}`
+        ];
+    }
+
+    rlpBytes(): Buffer {
+        return RLP.encode(this.toEncdoeObject());
+    }
+
+    fromBytes(bytes: Buffer): ExtensionMessage {
+        const decodedbytes = RLP.decode(bytes);
+        const version = decodedbytes[0].readUIntBE(0, 1);
+        const protocolId = decodedbytes[1].readUIntBE(0, 1);
+        const extensionName = decodedbytes[2];
+        const extensionVersion = decodedbytes[3];
+        const data = decodedbytes[4];
+
+        if (protocolId === MessageType.ENCRYPTED_ID) {
+            return new ExtensionMessage(
+                version,
+                extensionName,
+                extensionVersion,
+                { type: "encrypted", data }
+            );
+        } else if (protocolId === MessageType.UNENCRYPTED_ID) {
```

Comment:
It is better not to use `else`, if the preceding block always returns.

Reply:
I resolved it.